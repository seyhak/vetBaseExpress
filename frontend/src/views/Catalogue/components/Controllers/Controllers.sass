.controllers-wrapper
  display: flex
  padding: 1em
  margin-bottom: 1em
  background: white
  justify-content: space-between

  .search-status-wrapper
    display: flex
    justify-content: flex-start
    align-items: center
