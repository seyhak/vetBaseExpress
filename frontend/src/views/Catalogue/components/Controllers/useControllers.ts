import { useState, useEffect, useCallback } from "react"
import { useLocation, useNavigate } from "react-router-dom"
import { ViewsPaths } from "types/routes.types"

export const useControllers = () => {
  const location = useLocation()
  const navigate = useNavigate()
  const [search, setSearch] = useState<null | string>(null)

  useEffect(() => {
    const newSearch =
      Object.fromEntries(new URLSearchParams(location.search))?.search || null
    setSearch(newSearch)
  }, [location])

  const onResetSearchClick = useCallback(() => {
    navigate({
      pathname: ViewsPaths.Catalogue,
    })
  }, [navigate])

  return {
    search,
    onResetSearchClick,
  }
}
