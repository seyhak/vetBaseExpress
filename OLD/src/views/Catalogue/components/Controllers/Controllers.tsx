import {
  Box,
  ButtonGroup,
  IconButton,
  Typography,
  Tooltip,
} from "@mui/material"
import { ConfirmDeleteButtonWithDialog } from "../ConfirmDeleteButtonWithDialog/ConfirmDeleteButtonWithDialog"

import AddCircleOutlineIcon from "@mui/icons-material/AddCircleOutline"
import ModeEditIcon from "@mui/icons-material/ModeEdit"
import SaveIcon from "@mui/icons-material/Save"
import RestartAltIcon from "@mui/icons-material/RestartAlt"
import { useControllers } from "./useControllers"
import "./Controllers.sass"

export type ControllersProps = {
  isItemSelected: boolean
  isEditModeOn: boolean
  onAddClick: () => void
  onSaveClick: () => Promise<void>
  onEditClick: () => void
  onDeleteClick: () => Promise<void>
}
export const Controllers = ({
  isItemSelected,
  isEditModeOn,
  onAddClick,
  onSaveClick,
  onEditClick,
  onDeleteClick,
}: ControllersProps) => {
  const { search, onResetSearchClick } = useControllers()

  return (
    <Box className="controllers-wrapper">
      <Box className="search-status-wrapper">
        {!!search && (
          <Typography variant="h6">Results for "{search}"</Typography>
        )}
      </Box>
      <ButtonGroup
        variant="outlined"
        aria-label="outlined primary button group"
      >
        {!!search && (
          <Tooltip title="Reset search">
            <IconButton
              aria-label="reset-search"
              color="primary"
              onClick={onResetSearchClick}
            >
              <RestartAltIcon />
            </IconButton>
          </Tooltip>
        )}
        <Tooltip title="Add position">
          <IconButton aria-label="add" color="primary" onClick={onAddClick}>
            <AddCircleOutlineIcon />
          </IconButton>
        </Tooltip>
        {isItemSelected && (
          <>
            {isEditModeOn && (
              <IconButton
                onClick={onSaveClick}
                aria-label="save"
                color="primary"
              >
                <SaveIcon />
              </IconButton>
            )}
            <Tooltip title="Edit position">
              <IconButton
                onClick={onEditClick}
                aria-label="edit"
                color="primary"
              >
                <ModeEditIcon />
              </IconButton>
            </Tooltip>
            <ConfirmDeleteButtonWithDialog onDeleteClick={onDeleteClick} />
          </>
        )}
      </ButtonGroup>
    </Box>
  )
}
