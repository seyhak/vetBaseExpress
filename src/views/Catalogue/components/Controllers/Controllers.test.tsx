import { fireEvent, render, screen, waitFor } from "@testing-library/react"
import { MemoryRouter } from "react-router-dom"
import { Controllers, ControllersProps } from "./Controllers"

describe("Controllers", () => {
  let props: ControllersProps
  beforeEach(() => {
    props = {
      isItemSelected: false,
      isEditModeOn: false,
      onAddClick: jest.fn(),
      onSaveClick: jest.fn(),
      onEditClick: jest.fn(),
      onDeleteClick: jest.fn(),
    }
  })
  test.each([
    ["onAddClick", "add"],
    ["onSaveClick", "save"],
    ["onEditClick", "edit"],
  ])("no search, buttons %s", async (funcName, label) => {
    props.isItemSelected = true
    props.isEditModeOn = true

    render(
      <MemoryRouter>
        <Controllers {...props} />
      </MemoryRouter>,
    )

    const btn = screen.getByLabelText(label)
    await fireEvent.click(btn)
    const func = props[funcName as keyof ControllersProps]
    await waitFor(() => {
      expect(func).toBeCalled()
    })
  })
  test("no search behavior", () => {
    const view = render(
      <MemoryRouter>
        <Controllers {...props} />
      </MemoryRouter>,
    )

    expect(view.asFragment()).toMatchSnapshot()
    expect(screen.queryByText("Results for")).toBeFalsy()
    expect(screen.queryByLabelText("reset-search")).toBeFalsy()
  })
  test("search behavior", () => {
    render(
      <MemoryRouter initialEntries={["/list?search=banan"]}>
        <Controllers {...props} />
      </MemoryRouter>,
    )

    expect(screen.getByText('Results for "banan"')).toBeInTheDocument()
    expect(screen.getByLabelText("reset-search")).toBeInTheDocument()
  })
  test("no item selected behavior", () => {
    render(
      <MemoryRouter>
        <Controllers {...props} />
      </MemoryRouter>,
    )

    expect(screen.getByLabelText("add")).toBeInTheDocument()
    expect(screen.queryByLabelText("save")).toBeFalsy()
    expect(screen.queryByLabelText("edit")).toBeFalsy()
    expect(screen.queryByLabelText("delete")).toBeFalsy()
  })
  test("item selected behavior not in editmode", () => {
    props.isItemSelected = true

    render(
      <MemoryRouter>
        <Controllers {...props} />
      </MemoryRouter>,
    )

    expect(screen.getByLabelText("add")).toBeInTheDocument()
    expect(screen.getByLabelText("edit")).toBeInTheDocument()
    expect(screen.getByLabelText("delete")).toBeInTheDocument()
    expect(screen.queryByLabelText("save")).toBeFalsy()
  })
  test("item selected behavior, is in editmode", () => {
    props.isItemSelected = true
    props.isEditModeOn = true

    render(
      <MemoryRouter>
        <Controllers {...props} />
      </MemoryRouter>,
    )

    expect(screen.getByLabelText("add")).toBeInTheDocument()
    expect(screen.getByLabelText("edit")).toBeInTheDocument()
    expect(screen.getByLabelText("delete")).toBeInTheDocument()
    expect(screen.getByLabelText("save")).toBeInTheDocument()
  })
})
