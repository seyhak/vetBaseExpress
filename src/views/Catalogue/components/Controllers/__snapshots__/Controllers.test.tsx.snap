// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controllers no search behavior 1`] = `
<DocumentFragment>
  <div
    class="controllers-wrapper MuiBox-root css-0"
  >
    <div
      class="search-status-wrapper MuiBox-root css-0"
    />
    <div
      aria-label="outlined primary button group"
      class="MuiButtonGroup-root MuiButtonGroup-outlined css-edsu1d-MuiButtonGroup-root"
      role="group"
    >
      <button
        aria-label="add"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeMedium css-1kuq5xv-MuiButtonBase-root-MuiIconButton-root"
        data-mui-internal-clone-element="true"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="AddCircleOutlineIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4V7zm-1-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
